oeps/oep-0039-yaml-config.rst

```diff
@@ -0,0 +1,146 @@
+forces you to have to specify --settings= everywhere you interface with the application(management commands nginx, etc.) These files also often modify the config after they are loaded by add/remove/appending to settings when the file is executed.
+This interpreting of the configuration at runtime in python makes the configuration harder to reason about and does not provide additional functionality. Since many values must be defined for the application to function these files will often either load other configuration files and then override settings(adding another layer of configuration indirection), or redefine all the settings defined in the other modules, typically only changing a few values and increasing code duplication.  eg. In the platform the plugins are forced to know all of the environments that the application code might have and to provide boilerplate code with magic strings in it to get their configuration loaded.  All of this boilerplate code could be removed using this method.
+
+Having all changes in the yaml file makes it so that all plugin code can simply assume its settings are defined in the yaml file and therefore the python settings module without having to worry about defining configuration for various environments.
+This eliminates a lot of boilerplate code for plug in developers.
+
+Starting from the production defaults and making changes for development minimizes the differences in configuration between production and development environments.
+
+ENV variables should not be used to hold secrets as they are often leaked by web frameworks and plug ins in development configurations, and are accessed much easier by an attacker than a file in a variety of scenarios.
+
+As we make more django applications it is easier to maintain them if they are configured the same way.
+
+defaults.py provides a place where one can expect to find documentation about each available setting, if that documentation exists.
+
+Backward Compatibility
+======================
+
+This change is not backwards compatible, but it need not be implemented in each application immediately.
+It is meant to serve as a set of best practices for moving forward, but
+
+In general the migration path for old applications would look like this:
+
+	For all applications:
+    rename production.py to settings.py
+    rename base.py or common.py to defaults.py
+    make production.py source defaults.py and read yaml into python settings module 
+    (it should not specify every key like it does in many apps, it should load keys from the yaml into keys in the settings module without knowledge of what keys are possible.)
+    
+    Make yaml files for overrides in any other environments and remove the additional python environments.
+    Remove any --settings= from places where we interface with the application such as management commands, and from documentation for the application.
```

Comment:
How should devs expect to set the right ENV variable when switching between dev, test, etc.  Will we capture this in a `Makefile` and set the right env for you?  Will they need to set it on their own?

Reply:
I will clarify - my intent was that there should only be one python django configuration environment and that environment should load a yml, changing configuration would always be done in the yml, if you had multiple environments to configure you would likely have multiple ymls.

Reply:
There are 14 of these environments in the edx platform lms presently. (similar number in the cms)